hail/src/main/scala/is/hail/expr/ir/Requiredness.scala

```diff
@@ -0,0 +1,422 @@
+        requiredness.union(sReq.required && sReq.children.forall(_.required))
+      case NDArrayConcat(nds, axis) =>
+        val ndsReq = lookupAs[RIterable](nds)
+        requiredness.unionFrom(ndsReq.elementType)
+        requiredness.union(ndsReq.required)
+      case NDArrayRef(nd, idxs) =>
+        val ndReq = lookupAs[RNDArray](nd)
+        requiredness.unionFrom(ndReq.elementType)
+        requiredness.union(ndReq.required && idxs.forall(lookup(_).required))
+      case NDArraySlice(nd, slices) =>
+        val slicesReq = lookupAs[RTuple](slices)
+        val allSlicesRequired = slicesReq.types.forall {
+          case r: RTuple => r.required && r.types.forall(_.required)
+          case r => r.required
+        }
+        requiredness.unionFrom(lookup(nd))
+        requiredness.union(slicesReq.required && allSlicesRequired)
+      case NDArrayFilter(nd, keep) =>
+        requiredness.unionFrom(lookup(nd))
+      case NDArrayMap(nd, name, body) =>
+        requiredness.union(lookup(nd).required)
+        coerce[RNDArray](requiredness).unionElement(lookup(body))
+      case NDArrayMap2(l, r, _, _, body) =>
+        requiredness.union(lookup(l).required && lookup(r).required)
+        coerce[RNDArray](requiredness).unionElement(lookup(body))
+      case NDArrayMatMul(l, r) =>
+        requiredness.unionFrom(lookup(l))
+        requiredness.union(lookup(r).required)
+      case NDArrayQR(nd, mode) => requiredness.maximize()
+      case MakeStruct(fields) =>
+        fields.foreach { case (n, f) =>
+          coerce[RStruct](requiredness).field(n).unionFrom(lookup(f))
+        }
+      case MakeTuple(fields) =>
+        fields.foreach { case (i, f) =>
+          coerce[RTuple](requiredness).types(i).unionFrom(lookup(f))
+        }
+      case SelectFields(old, fields) =>
+        val oldReq = lookupAs[RStruct](old)
+        requiredness.union(oldReq.required)
+        fields.foreach { n =>
+          coerce[RStruct](requiredness).field(n).unionFrom(oldReq.field(n))
+        }
+      case InsertFields(old, fields, _) =>
+        lookup(old) match {
+          case oldReq: RStruct =>
+            requiredness.union(oldReq.required)
+            val fieldMap = fields.toMap.mapValues(lookup)
+            coerce[RStruct](requiredness).fields.foreach { f =>
+              f.typ.unionFrom(fieldMap.getOrElse(f.name, oldReq.field(f.name)))
+            }
+          case _ => fields.foreach { case (n, f) =>
+            coerce[RStruct](requiredness).field(n).unionFrom(lookup(f))
+          }
+        }
+      case GetField(o, name) =>
+        val oldReq = lookupAs[RStruct](o)
+        requiredness.union(oldReq.required)
+        requiredness.unionFrom(oldReq.field(name))
+      case GetTupleElement(o, idx) =>
+        val oldReq = lookupAs[RTuple](o)
+        requiredness.union(oldReq.required)
+        requiredness.unionFrom(oldReq.fields(idx).typ)
+      case x: ApplyIR => requiredness.unionFrom(lookup(x.explicitNode))
```

Comment:
Prefer to separate this into something managed by the Apply rather than using explicitNode.